File: 1000-9999/1219.py
'''
1219번
오민식의 고민
'''

import sys
from collections import deque
input=sys.stdin.readline
INF=987654321

def check(i):
    global graph, dist
    q=deque([i]); visited={}
    while q:
        node=q.pop()
        if node==e:
            return True
        visited[node]=True
        for nxt, cost in graph[node]:
            if nxt not in visited:
                q.append(nxt)
    return False

def solve():
    global graph, dist
    dist[s]=0
    for _ in range(n-1):
        for i in range(n):
            for nxt, cost in graph[i]:
                if dist[i]==INF: continue
                if dist[nxt]>dist[i]+cost:
                    dist[nxt]=dist[i]+cost
    tmp=dist[e]
    if tmp==INF: 
        print("gg")
        return
    for i in range(n):
        for nxt, cost in graph[i]:
            if dist[i]==INF: continue
            if dist[nxt]>dist[i]+cost:
                dist[nxt]=dist[i]+cost
                if check(nxt):
                    print("Gee")
                    return
    print(-dist[e]+vil[s])

graph={}
n, s, e, m=map(int, input().split())
for _ in range(m):
    a, b, c=map(int, input().split())
    if a not in graph: graph[a]=[]
    graph[a].append([b, c])
vil=list(map(int, input().split()))

for i in range(n):
    if i not in graph: graph[i]=[]
    for g in graph[i]:
        g[1]-=vil[g[0]]
dist=[INF]*n

solve()

File: 1000-9999/1219.cpp
//1219번 오민식의 고민

#include <bits/stdc++.h>
using namespace std;
#define INF 987654321

int n, m, s, e, a, b, c;
int cost[101][101], dist[101], vil[101], visited[101];

int check(int j){
    queue<int> q;
    q.push(j);
    for(int i=0; i<100; i++) visited[i]=0;
    while(!q.empty()){
        int node=q.front(); q.pop();
        if(node==e) return true;
        visited[node]=1;
        for(int i=0; i<n; i++){
            if(cost[node][i]!=INF && !visited[i]) q.push(i);
        }
    }
    return false;
}

int main(void){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    
    cin>>n>>s>>e>>m;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++) cost[i][j]=INF;
    }
    for(int i=0; i<m; i++){
        cin>>a>>b>>c;
        cost[a][b]=min(cost[a][b], c);
    }
    for(int i=0; i<n; i++){
        cin>>vil[i];
        for(int j=0; j<n; j++){
            if(cost[j][i]!=INF) cost[j][i]-=vil[i];
        }
    }

    for(int i=0; i<n; i++) dist[i]=INF;
    dist[s]=-vil[s];
    for(int i=0; i<n-1; i++){
        for(int j=0; j<n; j++){
            if(dist[j]==INF) continue;
            for(int k=0; k<n; k++){
                if(cost[j][k]!=INF && dist[k]>cost[j][k]+dist[j])
                    dist[k]=cost[j][k]+dist[j];
            }
        }
    }
    int tmp=dist[e];
    if(tmp==INF){
        cout<<"gg"<<"\n";
        return 0;
    }
    if(dist[e]!=INF)
        for(int i=0; i<n; i++){
            if(dist[i]==INF) continue;
            for(int j=0; j<n; j++){
                if(cost[i][j]!=INF && dist[j]>cost[i][j]+dist[i]){
                    dist[j]=cost[i][j]+dist[i];
                    if(check(j)){
                        cout<<"Gee"<<"\n";
                        return 0;
                    }
                }
            }
        } 

    cout<<-dist[e]<<"\n";
}